<a href="https://colab.research.google.com/github/OwaisAbuSalah/Prediction-of-Product-Sales/blob/main/Data_Cleaning_Core_Starter_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Cleaning Core [Starter Notebook]



- Student Name:

## The Data

You will be working with a modified version of the Titanitc dataset:

- [(Modified) Titanic Data](https://drive.google.com/file/d/1Uw7RzQMYSARCFnzJQqqNlZpUouXr16-8/view?usp=sharing)

- [Data Dictionary](https://docs.google.com/document/d/1uEeANvKURYmu-HUDGzbGlMmvAUwooatUg-gHUCDMh-Y/edit?usp=sharing)

___

- Save the files to your Course 1 > Week 2 > Data Folder.


## A) Load and Preview Data


- Mount google drive and navigate to the file on the Files sidebar.


In [13]:
# Mount google drive
from google.colab import drive
drive.mount('/content/drive')

MessageError: Error: credential propagation was unsuccessful


- Navigate to the file on the Files side bar and copy the path
- Paste it below and save it as a variable called "fpath"

In [14]:
# Import required packages
import pandas as pd
fpath="/content/drive/MyDrive/AXSOSACADEMY/01-Fundamentals/Week02/Data/titanic_dirty.csv"

# Save the filepath and load the data as "df_titanic"
df_taitanic = pd.read_csv(fpath)



FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/AXSOSACADEMY/01-Fundamentals/Week02/Data/titanic_dirty.csv'

In [15]:
# Display the first 5 rows
df_taitanic.head()


NameError: name 'df_taitanic' is not defined

In [94]:
# Display the last 5 rows
df_taitanic.tail()


Unnamed: 0  PassengerId  Survived  Pclass  \
901         622          623         1       3   
902          77           78         0       3   
903         715          716         0       3   
904         623          624         0       3   
905         888          889         0       3   

                                           Name     Sex   Age SibSp  Parch  \
901                            Nakid, Mr. Sahid    male  20.0     1      1   
902                    Moutal, Mr. Rahamin Haim    male   NaN     0      0   
903  Soholt, Mr. Peter Andreas Lauritz Andersen    male  19.0     0      0   
904                 Hansen, Mr. Henry Damsgaard    male  21.0     0      0   
905    Johnston, Miss. Catherine Helen "Carrie"  female   NaN     1      2   

         Ticket    Fare  Cabin Embarked  
901        2653  $15.74    NaN        C  
902      374746   $8.05    NaN        S  
903      348124   $7.65  F G73        S  
904      350029   $7.85    NaN        S  
905  W./C. 6607  $23.45    NaN        S

- Display summary information (including dtypes, number of rows, number of columns).


In [95]:
# Display the info for the dataframe
df_taitanic.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 906 entries, 0 to 905
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   906 non-null    int64  
 1   PassengerId  906 non-null    int64  
 2   Survived     906 non-null    int64  
 3   Pclass       906 non-null    int64  
 4   Name         906 non-null    object 
 5   Sex          906 non-null    object 
 6   Age          723 non-null    float64
 7   SibSp        906 non-null    object 
 8   Parch        906 non-null    int64  
 9   Ticket       906 non-null    object 
 10  Fare         906 non-null    object 
 11  Cabin        207 non-null    object 
 12  Embarked     904 non-null    object 
dtypes: float64(1), int64(5), object(7)
memory usage: 92.1+ KB


**Scan the .info() to answer the following questions about the data:**

>Add your answers in this Markdown cell, right below each question:

**Q: A1) How many rows are there?**
- Your Answer:
  - `906 rows`

**Q: A2) How many columns are there?**
- Your Answer:
  - `13 columns`




**Q: A3)How many numeric columns?**
- Your Answer:
  - `6 columns`

**Q: A4) How many categorical/string columns are there (dtype='object')?**
- Your Answer:
  - `7 rows`

 B) Check Data Types and Convert (if required)


> Notice there are several columns with "object" datatype.

- **Use select_dtypes to view only the object columns**


In [7]:
# Use .select_dtypes to view only object columns
df_taitanic = df_taitanic.select_dtypes(include=['object'])
df_taitanic.head()


NameError: name 'df_taitanic' is not defined

- Inspect the values in each column to confirm if they should be converted to numeric dtypes and answer the following questions:




**Q: B1) Which columns should be converted to a numeric dtype?**  (Hint: there are 2 columns that need to be converted)
- **Your Answer:**
  - `SibSp`
  - `Fare`
    




### Convert Dtypes

#### Convert Fare

In [99]:
# Inspect the Fare column's value counts
df_taitanic = df_taitanic["Fare"].value_counts(dropna=False)


KeyError: 'Fare'



>In a Markdown cell below your displayed info, answer the following questions:

**Q: B2)  Do you see any values that are string characters that need to be replaced?**
- Your Answer:
  - `$`

**Q: B3) If so, do you need to replace a single part of a string (`df[col].str.replace`) or an entire value (`df[col].replace`)?**
- Your Answer:
  - `df[col].str.replace`

**Perform the necessary steps to convert to numeric dtype:**

- 1) Replace/remove the non-numeric values.

In [75]:
# Remove  the "$" from fare column by replacing it with nothing
df_taitanic = df_taitanic['Fare'].str.replace('$','')
# Check the .value_counts to confirm the change
df_taitanic.value_counts('Fare')

KeyError: 'Fare'


- 2) Convert the column to a float dtype.

In [73]:
# Change the datatype of Fare column to float
df_taitanic = df_taitanic['Fare'].astype(float)
# Confirm the datatype
df_taitanic['Fare'].dtype

KeyError: 'Fare'

#### Convert SibSp

In [53]:
# Inspect the Sibsp column's value counts
df_taitanic["SibSp"].value_counts(dropna=False)

SibSp
0      618
1      212
2       27
4       18
3       16
8        7
5        5
one      3
Name: count, dtype: int64

>In a Markdown cell below your displayed info, answer the following questions:

**Q: B3)  Do you see any values that are string characters that need to be replaced?**
- Your Answer:
  - `one`

**Q: B4) If so, do you need to replace a single part of a string (`df[col].str.replace`) or an entire value (`df[col].replace`)?**
- Your Answer:
  - `df[col].str.replace`

**Perform the necessary steps to convert to numeric dtype:**

- 1) Replace/remove the non-numeric values.

In [54]:
# There are 3 values of "one" which should be replaced with "1"
df_taitanic['SibSp'] = df_taitanic['SibSp'].str.replace('one','1')
# Display the value counts again to confirm the change
df_taitanic["SibSp"].value_counts(dropna=False)

SibSp
0    618
1    215
2     27
4     18
3     16
8      7
5      5
Name: count, dtype: int64


- 2) Convert the column to a float dtype.

In [55]:
# Convert the SibSp column to integer dtype
df_taitanic['SibSp'] = df_taitanic['SibSp'].astype(float)
# Confirm the datatype
df_taitanic['SibSp'].dtype

dtype('float64')

## C) Drop Unwanted Columns

- Drop the 'Unnamed: 0' column because it is an error

In [56]:
# Drop the Unnamed: 0 column
df_taitanic = df_taitanic.drop(columns=['Unnamed: 0'])
# Display the .head to confirm the change
df_taitanic.head()

KeyError: "['Unnamed: 0'] not found in axis"

## D) Rename Columns
(that have Confusing/Abbreviated Column Names)

In the cell below:
- Rename the SibSp column as "Siblings or Spouses"
- Rename the Parch as "Parents or Children"

In [ ]:
# Create a dictionary called rename_dict
# with current column names as keys
# and new replacement names as the values.


In [ ]:
# Use the rename_dict to change the column names in the dataframe

# Check the .head() to confirm the nams changed.


## E) Duplicate Rows

### Identifying Duplicates


- Use `df.duplicated` to identify if there are any duplicate rows in the data.


In [ ]:
# Save a filter called "duplicated_rows" that identifies duplicated rows


# Get the sum of duplicated_rows filter to see the number of duplicate rows



**Q: E1) How many duplicate rows are in the data?**
- Your Answer:
  - `...`

  
- Visualize all copies of the duplicate rows using df.duplicated(keep=False).
  - Sort the rows by the PassengerID column to view duplicate rows together.

In [ ]:
# Save a new filter called "duplicated_rows_all" that identifies ALL
# of the duplicated rows (including the first). Hint: use the "keep" argument.


# Calculate the sum of the duplicated_rows_all filter


In [ ]:
# View all duplicated rows using your duplicated_rows_all filter
# Sort by PassengerId


In [ ]:
# Drop duplicates (but keep the first!)

# Obtain the sum of duplicated to confirm all have been dropped


## F) Unique Identifiers
- Find and Drop Columns with Unique Values for Every Row.

In [ ]:
# Use .nunqiue() to get # of unique values for each column and save as a variable.


In [ ]:
# Use the result to determine if any of the columns have 100% unique values.


**Q: F1) Which 2 features have 100% unique values in every row?**
- Your Answer:
  - Column 1: `...`
  - Column 2: `...`




>Generally, a column that is NOT the index, and has a unique value for each row will not be helpful for analysis and should be dropped from the data.

- Inspect the two columns with 100% unique values:


In [ ]:
# Slice the two columns with 100% unique values to view



- **Of the two columns with 100% unique values:**
  - Set the numeric column as the DataFrame's index.
  - Drop the string column from the DataFrame.


In [ ]:
# Set the numeric column as the DataFrame's index.


In [ ]:
# Drop the string column from the DataFrame.


___

## G) Missing Values

In [ ]:
# Get the number of null values for all columns and save as a variable


In [ ]:
# Use this result to calculate the percentage of null values for each column.


In [ ]:
# Visualize the missing values using the missingno package's matrix plot.


#### Address the null values:

- Fill the null values in categorical columns with a placeholder category called MISSING.


In [ ]:
# save list of categorical column name. (Hint: select_dtypes)


In [ ]:
# use the list of cat columns to fill null values with MISSING



- Fill the null values in numeric columns with a placeholder value of -1



In [ ]:
# save list of numeric column names (Hint: select_dtypes)


In [ ]:
# use the list of num columns to fill null values with -1


- Use either missingno's matrix plot or Pandas to confirm that there are 0 null values remaining.


In [ ]:
# Confirm no more nulls using Pandas


In [ ]:
# Confirm no more nulls using using missingno


## H) Consistent Values

- Check each string column's value counts for inconsistent values.
  - E.g. "Cat" vs "cat", "Y" vs. "Yes"



- Fix/replace the inconsistent values with the values you believe are correct.


- Display the value_counts again to confirm the values are now consistent.
  

In [ ]:
# Save a list of object columns (Hint: select_dtypes)


In [ ]:
# Check the nunique for just the object cols


- Note that out of the 4 categorical columns remaining:
  - 2 columns have a small number of categories  (Sex and Embarked)
  - 2 columns have over 100 categories (Ticket and Cabin).

- We do not want to keep categorical features with so many categories.


- **Drop the Ticket column and Cabin column.**

In [ ]:
# Drop the columns listed above


### For Each Remaining Categorical Column:

- Display the value counts for the column
- Take a note of any values that are not consistent.
- Remove/fix the inconsistent values
- Confirm by checking the value counts again

In [ ]:
# Remake the list of string columns (after dropping previous cols)

# loop through the list of string columns

  # print the value counts for the column

  # Print an empty line for readability


**Q: H1) Which columns have inconsistent values that need to be addressed?**
- Your Answer:
  - `...`
  - `...`

### Checking & Fixing Consistency: Column 1

In [ ]:
# Check for inconsistencies by inspecting the value_counts for Embarked


In [ ]:
# Replace "Queenstown" with "Q"

# Check the value counts again to confirm


### Checking and Fixing Consistency: Column 2

In [ ]:
# # Check for inconsistencies by inspecting the values_counts for Sex


In [ ]:
# Replace "F" with "female"


# Check the value counts again to confirm


## I) Save the Final DataFrame to Csv

- Save the cleaned DataFrame to your Google Drive in the CodingDojo>  01-Fundamentals> Week02>Data folder.
  - Name the file "data-cleaning-core.csv"


In [ ]:
# Save the folder + filename as a variable called fpath


# Use to_csv to save the dataframe using the filename


- Confirm the data was saved correctly.

In [ ]:
# Read in your new .csv file with Pandas.

# Display the info summary.

# Display the first 5 rows.


### Save and Submit your work

When the assignment is complete, download your notebook as a .ipynb file and submit it on the learning platform.  